packages/react-cosmos-playground/src/components/ResponsiveLoader/index.js

```diff
@@ -0,0 +1,183 @@
+// @flow
+
+import React from 'react';
+import Header from './Header';
+import classNames from 'classnames';
+import localForage from 'localforage';
+import styles from './index.less';
+
+type Props = {
+  inputRef: (node: ?HTMLElement) => void,
+  src: string,
+  showResponsiveControls: boolean,
+  onFixtureUpdate: any,
+  devices: Array<{| label: string, width: number, height: number |}>,
+  fixture: Object
+};
+
+type State = {
+  containerWidth: number,
+  containerHeight: number,
+  savedWidth: ?number,
+  savedHeight: ?number
+};
+
+const RESPONSIVE_FIXTURE_WIDTH = '__cosmos__responsive-fixture-width';
+const RESPONSIVE_FIXTURE_HEIGHT = '__cosmos__respinsive-fixture-height';
+
+const PADDING = 16;
+const BORDER_WIDTH = 2;
+
+class ResponsiveLoader extends React.Component<Props, State> {
+  state = {
+    containerWidth: 10000,
+    containerHeight: 10000,
+    savedWidth: null,
+    savedHeight: null
+  };
+  scalableDiv: ?HTMLElement;
+
+  async componentDidMount() {
+    window.addEventListener('resize', this.updateContainerWidth);
+    setTimeout(this.updateContainerWidth, 1000);
+
+    const [savedWidth, savedHeight] = await Promise.all([
+      localForage.getItem(RESPONSIVE_FIXTURE_WIDTH),
+      localForage.getItem(RESPONSIVE_FIXTURE_HEIGHT)
+    ]);
+
+    this.setState({ savedWidth, savedHeight });
+  }
+
+  componentWillUnmount() {
+    window.removeEventListener('resize', this.updateContainerWidth);
+  }
+
+  updateDimensions = (width: number, height: number) => {
+    const { onFixtureUpdate } = this.props;
+    localForage.setItem(RESPONSIVE_FIXTURE_WIDTH, width);
+    localForage.setItem(RESPONSIVE_FIXTURE_HEIGHT, height);
+    onFixtureUpdate({ viewport: { width, height } });
+    this.setState({ savedWidth: width, savedHeight: height });
+  };
+
+  updateContainerWidth = () => {
+    if (!this.scalableDiv) {
+      return;
+    }
+    const nextContainerWidth = this.scalableDiv.getBoundingClientRect().width;
+    const nextContainerHeight = this.scalableDiv.getBoundingClientRect().height;
+    if (
+      nextContainerWidth !== this.state.containerWidth ||
+      nextContainerHeight !== this.state.containerHeight
+    ) {
+      this.setState({
+        containerWidth: nextContainerWidth - 2 * (PADDING + BORDER_WIDTH),
+        containerHeight: nextContainerHeight - 2 * (PADDING + BORDER_WIDTH)
+      });
+    }
+  };
+
+  render() {
+    const {
+      inputRef,
+      src,
+      devices,
+      showResponsiveControls,
+      fixture
+    } = this.props;
+    const { savedWidth, savedHeight } = this.state;
+
+    const { viewport = {} } = fixture;
+    const width =
+      viewport.width === 0 || viewport.width
+        ? viewport.width
+        : savedWidth || 320;
+    const height =
+      viewport.height === 0 || viewport.height
+        ? viewport.height
+        : savedHeight || 568;
+    const scale = true;
```

Comment:
It seems like scaling is an always-on option for future implementation. Not sure how long in the future you intend to make it variable, but maybe leave a comment for now?

Reply:
I'd definitely like to get to it before the feature is released. Have added a comment.